pkg/api/handlers/auth_handlers.go

```diff
@@ -0,0 +1,230 @@
+	// FIXME: How is loginPage meant to be used here?
+	loginPage := "/oversikt"
+
+	redirectURI, _ := ctx.Cookie(h.cookies.Redirect.Name)
+	if redirectURI != "" {
+		loginPage = redirectURI
+	}
+
+	if strings.HasPrefix(ctx.Request.Host, "localhost") {
+		loginPage = "http://localhost:8080" + loginPage
+	}
+
+	deleteCookie(ctx, h.cookies.Redirect.Name, h.cookies.Redirect.Path, h.cookies.Redirect.Domain)
+
+	code := ctx.Request.URL.Query().Get("code")
+	if len(code) == 0 {
+		return loginPage + "?error=unauthenticated", errors.New("unauthenticated")
+	}
+
+	oauthCookie, err := ctx.Cookie(h.cookies.OauthState.Name)
+	if err != nil {
+		h.log.Infof("Missing oauth state cookie: %v", err)
+		return loginPage + "?error=invalid-state", errors.New("invalid state")
+	}
+
+	deleteCookie(ctx, h.cookies.OauthState.Name, h.cookies.OauthState.Path, h.cookies.OauthState.Domain)
+
+	state := ctx.Request.URL.Query().Get("state")
+	if state != oauthCookie {
+		h.log.Info("Incoming state does not match local state")
+		return loginPage + "?error=invalid-state", errors.New("invalid state")
+	}
+
+	sess, err := h.authService.CreateSession(ctx.Request.Context(), code)
+	if err != nil {
+		h.log.WithError(err).Error("unable to create session")
+
+		return loginPage + "?error=unauthenticated", fmt.Errorf("unable to create session: %w", err)
+	}
+
+	ctx.SetCookie(
+		h.cookies.Session.Name,
+		sess.Token,
+		h.cookies.Session.MaxAge,
+		h.cookies.Session.Path,
+		h.cookies.Session.Domain,
+		h.cookies.Session.Secure,
+		h.cookies.Session.HttpOnly,
+	)
+
+	return loginPage, nil
+}
+
+func deleteCookie(ctx *gin.Context, name, host, path string) {
+	ctx.SetCookie(name, "", -1, path, host, true, true)
+}
+
+func (h *AuthHandler) LoginHandler(dryRun bool) gin.HandlerFunc {
+	return func(ctx *gin.Context) {
+		// FIXME: We should not propagate dryRun, but rather do this by injecting a
+		//  different service, e.g., a mock service, and by passing in a different
+		// cookie configuration, perhaps?
+		if dryRun {
+			if err := h.createDryRunSession(ctx); err != nil {
+				h.log.Error("creating dryrun session")
+			}
+
+			ctx.Redirect(http.StatusSeeOther, "http://localhost:8080/oversikt")
+
+			return
+		}
+
+		ctx.SetSameSite(h.cookies.Redirect.GetSameSite())
+		ctx.SetCookie(
+			h.cookies.Redirect.Name,
+			ctx.Request.URL.Query().Get("redirect_uri"),
+			h.cookies.Redirect.MaxAge,
+			h.cookies.Redirect.Path,
+			h.cookies.Redirect.Domain,
+			h.cookies.Redirect.Secure,
+			h.cookies.Redirect.HttpOnly,
+		)
+
+		oauthState := uuid.New().String()
+		ctx.SetSameSite(h.cookies.OauthState.GetSameSite())
+		ctx.SetCookie(
+			h.cookies.OauthState.Name,
+			oauthState,
+			h.cookies.OauthState.MaxAge,
+			h.cookies.OauthState.Path,
+			h.cookies.OauthState.Domain,
+			h.cookies.OauthState.Secure,
+			h.cookies.OauthState.HttpOnly,
+		)
+
+		ctx.Redirect(http.StatusSeeOther, h.authService.GetLoginURL(oauthState))
+	}
+}
+
+// FIXME: Need to get rid of this
```

Comment:
Antar denne kommentaren følger den forrige `FIXME`? For da burde vi heller kommentere det, eller nevne det i oppgaven vi lager.